streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java

```diff
@@ -0,0 +1,229 @@
+        cache.put(name, binaryKey.get(), entry);
+    }
+
+    @Override
+    public KeyValueIterator<Windowed<K>, AGG> fetch(final K key) {
+        return findSessionsToMerge(key, 0, Long.MAX_VALUE);
+    }
+
+
+    public String name() {
+        return bytesStore.name();
+    }
+
+    @SuppressWarnings("unchecked")
+    public void init(final ProcessorContext context, final StateStore root) {
+        bytesStore.init(context, root);
+        initInternal((InternalProcessorContext) context);
+    }
+
+    @SuppressWarnings("unchecked")
+    void initInternal(final InternalProcessorContext context) {
+        this.context = context;
+
+        if (keySerde == null) {
+            keySerde = (Serde<K>) context.keySerde();
+        }
+
+
+        this.serdes = (StateSerdes<Windowed<K>, AGG>) new StateSerdes<>(bytesStore.name(),
+                                                                              new SessionKeySerde<>(keySerde),
+                                                                              aggSerde == null ? context.valueSerde() : aggSerde);
+
+
+        this.name = context.taskId() + "-" + bytesStore.name();
+        this.cache = this.context.getCache();
+        cache.addDirtyEntryFlushListener(name, new ThreadCache.DirtyEntryFlushListener() {
+            @Override
+            public void apply(final List<ThreadCache.DirtyEntry> entries) {
+                for (ThreadCache.DirtyEntry entry : entries) {
+                    final Bytes binaryKey = entry.key();
+                    final RecordContext current = context.recordContext();
+                    context.setRecordContext(entry.recordContext());
+                    try {
+                        if (flushListener != null) {
+                            final Windowed<K> key = SessionKeyBinaryConverter.from(binaryKey.get(), keySerde.deserializer());
+                            final AGG newValue = serdes.valueFrom(entry.newValue());
+                            final AGG oldValue = fetchPrevious(binaryKey);
+                            if (!(newValue == null && oldValue == null)) {
+                                flushListener.apply(key,
+                                                    newValue == null ? null : newValue, oldValue);
+                            }
+
+                        }
+                        bytesStore.put(binaryKey, entry.newValue());
+                    } finally {
+                        context.setRecordContext(current);
+                    }
+                }
+            }
+        });
+
+    }
+
+    private AGG fetchPrevious(final Bytes key) {
+        final byte[] bytes = bytesStore.get(key);
+        if (bytes == null) {
+            return null;
+        }
+        return serdes.valueFrom(bytes);
+    }
+
+
+    public void flush() {
+        cache.flush(name);
+        bytesStore.flush();
+    }
+
+    public void close() {
+        bytesStore.close();
```

Comment:
Yeah. The closing of the cache was dependent on another PR that has now been merged. `flush` doesn't matter so much as we always `flush` and then `close`, but i'll add it anyway